<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [69]:
import pandas as pd
import psycopg2

In [70]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
from personal import personal


connection = psycopg2.connect(
    dbname=personal['DBNAME'],
    user=personal['USER'],
    host=personal['HOST'],
    password=personal['PASSWORD'],
    port=personal['PORT']
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [71]:
# текст запроса
query3_1 = f'''select
            'Количество вакансий в базе Vacancies ' query,
            count(v.name) cnt_vacancies
            from public.vacancies  v 
         '''

In [72]:
# результат запроса
df_vacancies = pd.read_sql_query(query3_1, connection)
df_vacancies

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\4112126327.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacancies = pd.read_sql_query(query3_1, connection)


query  cnt_vacancies
0  Количество вакансий в базе Vacancies           49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [73]:
# текст запроса

query3_2 = f'''select 
            'Количество работодателей в таблице employers' query,
            count(name) cnt_employers
            from public.employers 
            
         '''

In [74]:
# результат запроса
df_employers = pd.read_sql_query(query3_2, connection)
df_employers

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\874782552.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employers = pd.read_sql_query(query3_2, connection)


query  cnt_employers
0  Количество работодателей в таблице employers          23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [75]:
# текст запроса
query3_3 = f'''select 
            'Количество регионов в таблице areas' query,
            count(name) cnt_areas
            from public.areas 
            
         '''

In [76]:
# результат запроса
df_areas = pd.read_sql_query(query3_3, connection)
df_areas

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\512970876.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas = pd.read_sql_query(query3_3, connection)


query  cnt_areas
0  Количество регионов в таблице areas       1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [77]:
# текст запроса
query3_4 = f'''select 
            'Количество сфер деятельности в таблице industries' query,
            count(name) cnt_industries
            from public.industries 
            
         '''

In [78]:
# результат запроса
df_industries = pd.read_sql_query(query3_4, connection)
df_industries


C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\3115407811.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industries = pd.read_sql_query(query3_4, connection)


query  cnt_industries
0  Количество сфер деятельности в таблице industries             294

***

In [79]:
# выводы по предварительному анализу данных
print (f'В базе данных "project_sql", содержатся данный о:')
display(df_employers)
display(df_vacancies)
display(df_industries) 
display(df_areas)

В базе данных "project_sql", содержатся данный о:


query  cnt_employers
0  Количество работодателей в таблице employers          23501

query  cnt_vacancies
0  Количество вакансий в базе Vacancies           49197

query  cnt_industries
0  Количество сфер деятельности в таблице industries             294

query  cnt_areas
0  Количество регионов в таблице areas       1362

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [80]:
# текст запроса
query4_1 = f'''select 
                a.name area,
                count(v.id) cnt
                from public.vacancies v
                join public.areas a on a.id = v.area_id 
                group by  a.name
                order by 2 desc'''

In [81]:
# результат запроса
df4_1 = pd.read_sql_query(query4_1, connection)
df4_1.head(10)

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\4018430656.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_1 = pd.read_sql_query(query4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698
6  Нижний Новгород  1670
7           Казань  1415
8        Краснодар  1301
9           Самара  1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [82]:
# текст запроса
query4_2 = f'''select 
                'Количество вакансий у которых заполнено поле с зарплатой' query,
                count(v.name)
                from public.vacancies v
                where (v.salary_from is not NULL) or (v.salary_to is not NULL)
                order by 2 desc'''

In [83]:
# результат запроса
df4_2 = pd.read_sql_query(query4_2, connection)
df4_2


C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\3812690898.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_2 = pd.read_sql_query(query4_2, connection)


query  count
0  Количество вакансий у которых заполнено поле с...  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [84]:
# текст запроса
query4_3 = f'''select 
                v.name vacancy,
                avg(v.salary_from) average_salary_from,
                avg(v.salary_to) average_salary_to
                from public.vacancies v
                where (v.salary_from is not NULL) or (v.salary_to is not NULL)
                group by  v.name
                order by 2 desc
                
                '''

In [85]:
# результат запроса
df4_3 = pd.read_sql_query(query4_3, connection)
print('Average_salary_from: ',round(df4_3['average_salary_from'].mean(),0), 'Average_salary_to: ',round(df4_3['average_salary_to'].mean(),0))


C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\692108986.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_3 = pd.read_sql_query(query4_3, connection)


Average_salary_from:  74955.0 Average_salary_to:  112427.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [86]:
# текст запроса
query4_4 = f'''select 
                v.schedule schedule, 
                v.employment employment,
                count(v.id) cnt_vacancies           
                from public.vacancies v
                group by v.schedule, v.employment
                order by 3 desc
                '''

In [87]:
# результат запроса
df4_4 = pd.read_sql_query(query4_4, connection)
df4_4

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\1702959072.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_4 = pd.read_sql_query(query4_4, connection)


schedule           employment  cnt_vacancies
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [88]:
# текст запроса
query4_5 = f'''select 
                v.experience experience, 
                count(v.id) cnt_vacancies            
                from public.vacancies v
                group by v.experience
                order by 2 asc
                '''

In [89]:
# результат запроса
df4_5 = pd.read_sql_query(query4_5, connection)
df4_5

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\2246630829.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_5 = pd.read_sql_query(query4_5, connection)


experience  cnt_vacancies
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

- Топ 5 регионов с предоставлением вакансии города миллионники: Москва, Санкт-Петербург,Минск, Новосибирск, Алматы. 
- У 48,9% (24073) вакансий указано поле с ЗП.
- Топ 3 самых распространённых типов рабочего графика и типа трудоустройства: Полный день/Полная занятость - 35367 (71,8%), Удаленная работа/   Полная занятость - 7802 (15,8%), Удаленная работа/Частичная занятость - 1312 (2,6%).
- Топ 3 самых распространенных опыта работы: от 1года до 3 лет - 26152 (53%), от 3 до 6 лет - 14511 (29,4%), нет опыта - 7197 (14,6%). 


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [90]:
# текст запроса
query5_1 = f'''select 
                e.name employer,
                --v.name vacancy, 
                count(e.name) cnt_employers            
                from public.vacancies v
                join public.employers e on e.id = v.employer_id
                group by e.name
                order by cnt_employers desc
                limit 5
                '''


In [91]:
# результат запроса
df5_1 = pd.read_sql_query(query5_1, connection)
df5_1

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\2599465926.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_1 = pd.read_sql_query(query5_1, connection)


employer  cnt_employers
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [92]:
# текст запроса
query5_2 = f'''(select 
                a.name area,
                count(v.name) cnt_vacancies,
                count(distinct e.name) cnt_employers
                from public.areas a
                join public.vacancies v on a.id = v.area_id
                join public.employers e on e.id = v.employer_id
                where v.name is not null
                group by a.name
                order by 3 desc
                limit 10)
                union all
                (select 
                a.name area,    
                count(v.name) cnt_vacancies, 
                count(e.name) cnt_employers 
                from public.areas a
                left join public.employers e on a.id = e.area
                left join public.vacancies v on e.area = v.area_id
                where v.name is null
                group by a.name
                order by 3 desc
                limit 5 ) 
                
                '''


In [93]:
# результат запроса
df5_2 = pd.read_sql_query(query5_2, connection)
df5_2

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\2906167698.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_2 = pd.read_sql_query(query5_2, connection)


area  cnt_vacancies  cnt_employers
0               Москва           5333           2320
1      Санкт-Петербург           2851           1448
2                Минск           2112            851
3          Новосибирск           2006            802
4               Алматы           1892            798
5         Екатеринбург           1698            742
6      Нижний Новгород           1670            656
7               Казань           1415            609
8            Краснодар           1301            578
9       Ростов-на-Дону           1131            486
10              Россия              0            410
11           Казахстан              0            207
12  Московская область              0             75
13  Краснодарский край              0             19
14            Беларусь              0             18

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [94]:
# текст запроса
query5_3 = f'''select 
                e.name employer,
                count(distinct a.name) cnt_areas
                from public.vacancies v
                join public.employers e on e.id=v.employer_id
                join public.areas a on a.id=v.area_id
                group by e.name 
                order by cnt_areas desc
                limit 15
                '''


In [95]:
# результат запроса
df5_3 = pd.read_sql_query(query5_3, connection)
df5_3

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\573122365.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_3 = pd.read_sql_query(query5_3, connection)


employer  cnt_areas
0                       Яндекс        181
1                   Ростелеком        152
2                   Спецремонт        116
3       Поляков Денис Иванович         88
4                    ООО ЕФИН          71
5                   Совкомбанк         63
6                          МТС         55
7   ЭФКО, Управляющая компания         49
8                         КРОН         48
9                 Почта России         48
10                 ИК СИБИНТЕК         46
11                       MCORE         46
12             Сбербанк-Сервис         45
13                       ANCOR         44
14                  Первый Бит         43

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [96]:
# текст запроса
query5_4 = f'''select 
                'Количество работодателей, у которых не указана сфера деятельности' query,
                count(e.name) cnt_employers
                from public.employers_industries e_i
                full join public.employers e on e.id = e_i.employer_id
                where e_i.industry_id is null
                group by e_i.industry_id
                order by cnt_employers desc
                
                '''

In [97]:
# результат запроса

df5_4=pd.read_sql_query(query5_4, connection)
df5_4

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\1547154052.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_4=pd.read_sql_query(query5_4, connection)


query  cnt_employers
0  Количество работодателей, у которых не указана...           8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [98]:
# текст запроса
query5_5 = f'''select
                e.name employer,
                count(e_i.industry_id) cnt_industries
                from public.employers_industries e_i
                full join public.employers e on e.id = e_i.employer_id
                group by e.name
                having count(e_i.industry_id) = 4
                order by employer asc
                limit 3
                '''


In [99]:
# результат запроса
df5_5=pd.read_sql_query(query5_5, connection)
df5_5

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\1403204011.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_5=pd.read_sql_query(query5_5, connection)


employer  cnt_industries
0  101 Интернет               4
1      21vek.by               4
2          2ГИС               4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [100]:
# текст запроса
query6_1 = f'''select
                i.name industry,
                count(e.name) cnt_employers
                from public.employers_industries e_i
                full join public.employers e on e.id = e_i.employer_id
                full join public.industries i on i.id =e_i.industry_id
                where i.name like 'Разработка программного обеспечения'
                group by i.name
                '''


In [101]:
# результат запроса
df6_1=pd.read_sql_query(query6_1, connection)
df6_1

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\3168786654.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_1=pd.read_sql_query(query6_1, connection)


industry  cnt_employers
0  Разработка программного обеспечения           3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [102]:
# код для получения списка городов-милионников
import urllib.parse as req
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[1]['Город'])

In [103]:
# текст запроса
query5_7 = f'''select
                a.name area,
                count(v.name) cnt_vacancies
                from public.vacancies v
                full join public.areas a on a.id = v.area_id
                full join public.employers e on e.id =v.employer_id
                where e.name like 'Яндекс' and a.name in  {cities}
                group by a.name
                
                union all
                select
                'TOTAL',
                count(v.name) TOTAL
                from public.vacancies v
                full join public.areas a on a.id = v.area_id
                full join public.employers e on e.id =v.employer_id
                where  e.name like 'Яндекс' and a.name in {cities}
                order by cnt_vacancies asc
                '''


In [104]:
# результат запроса
df5_7 = pd.read_sql_query(query5_7,connection)
df5_7

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\2396624492.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_7 = pd.read_sql_query(query5_7,connection)


area  cnt_vacancies
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            TOTAL            485

***

### выводы по анализу работодателей
- ТОП 5 работодателей по количеству вакансии: 1) Яндекс - 1933, 2) Ростелеком	491, 3)	Тинькофф - 444, 4) СБЕР	- 428, 5) Газпром нефть	- 331.
- Разработкой программного обеспечения занимаются 3553 (15% от общего количества).
- ТОП 3 компании у которых указано 4 сферы деятельности: 101 Интернет, 21vek.by, 2ГИС.
- У 35.8% (8419) компаний не указана сфера деятельности.
- В Москве расположено наибольшее количество работодателей - 2320 и количество вакансий -	5333. 
- Топ 5 работодателя у которых представлены вакансии в наибольшее количестве регионов: 1) Яндекс - 181 регион , 2) Ростелеком	- 152 регион 3) Спецремонт - 116 регион, 4) Поляков Денис Иванович - 88 регион, 5) ООО ЕФИН - 71 регион.
- Компания Яндекс предоставляет свои вакансии в всех городах миллионниках России. Общее количество 485 вакансий. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [105]:
# текст запроса
query6_1 = f'''select
                'Вакансии имеющие отношение к данным' query,
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (lower(v.name) like '%data%') or (lower(v.name) like '%данн%')  
                '''


In [106]:
# результат запроса
df6_1 = pd.read_sql_query(query6_1, connection)
df6_1

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\1506378273.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_1 = pd.read_sql_query(query6_1, connection)


query  cnt_vacancies
0  Вакансии имеющие отношение к данным           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [107]:
# текст запроса
query6_2 = f'''select
                'Количество подходящих вакансий для Junior DS' query,
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                and (v.name ilike '%junior%' or v.experience ilike '%Нет опыта%' or v.employment ilike '%Стажировка%')
                
                '''



In [108]:
# результат запроса
df6_2 = pd.read_sql_query(query6_2, connection)
df6_2

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\4141456713.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_2 = pd.read_sql_query(query6_2, connection)


query  cnt_vacancies
0  Количество подходящих вакансий для Junior DS             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [109]:
# текст запроса
query6_3 = f'''select
                'Вакансий для DS (key_skills: SQL/postgres)' query,
                --v.key_skills,v.name
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                
                and (v.key_skills like '%SQL%' or v.key_skills ilike '%postgres%' )
                
                '''


In [110]:
# результат запроса
df6_3 = pd.read_sql_query(query6_3, connection)
df6_3

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\1016452198.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_3 = pd.read_sql_query(query6_3, connection)


query  cnt_vacancies
0  Вакансий для DS (key_skills: SQL/postgres)            201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [111]:
# текст запроса
query6_4 = f'''select
                'Вакансий для DS (key_skills: Python)' query,
                --v.key_skills,v.name
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                
                and (v.key_skills like '%Python%')
                '''

In [112]:
# результат запроса
df6_4 = pd.read_sql_query(query6_4, connection)
df6_4

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\1720163166.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_4 = pd.read_sql_query(query6_4, connection)


query  cnt_vacancies
0  Вакансий для DS (key_skills: Python)            351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [113]:
# текст запроса
query6_5 = f'''select
                'Среднее значение ключевых навыков в вакансиях для DS' query,
                round(avg(length(v.key_skills) - length(replace(v.key_skills, '\t', '')) +1), 2) average
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                
                
                '''


In [114]:
# результат запроса
df6_5 = pd.read_sql_query(query6_5, connection)
df6_5

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\2816704726.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_5 = pd.read_sql_query(query6_5, connection)


query  average
0  Среднее значение ключевых навыков в вакансиях ...     6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [115]:
# текст запроса
query6_6 = f'''select
                'Средняя ЗП DS с опытом работы от 3 до 6 лет' query,
                round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to)),0)  average_salary
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%'))
                and v.experience ilike 'от 3 до 6 лет' 
                '''


In [116]:
# результат запроса
df6_6 = pd.read_sql_query(query6_6, connection)
df6_6

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\3610576656.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_6 = pd.read_sql_query(query6_6, connection)


query  average_salary
0  Средняя ЗП DS с опытом работы от 3 до 6 лет        243115.0

***

### выводы по предметному анализу

- Количество вакансий имеющих отношения к данным: 1771 (3,5% от общего количества)
- Среди них них на должность DataScience уровня junior: 51 вакансия.
- Количество вакансий на должность DataScience со знаниями (SQl/postgres): 201, со знаниями Python: 351 вакансия.
- Среднее количество необходимых наввыеков для трудоустройства на долность DataScience: 6,41
- Средняя заработная плата Data Science с опытом рабОты ОТ 3 ДО 6 ЛЕТ: 243 тыс. рублей. 

# Общий вывод по проекту

In [117]:

# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

На какую зарплату в среднем может рассчитывать дата-сайентист в зависимости от опыта работы?

In [118]:
# текст запроса
query7_1 = f'''select
                v.experience experience,
                round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to)),0)  average_salary
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%'))
                group by v.experience
                '''

In [119]:
# результат запроса
df7_1 = pd.read_sql_query(query7_1, connection)
df7_1

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\1635516678.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df7_1 = pd.read_sql_query(query7_1, connection)


experience  average_salary
0         Более 6 лет             NaN
1           Нет опыта         74643.0
2  От 1 года до 3 лет        139675.0
3       От 3 до 6 лет        243115.0

На какую зарплату в среднем может рассчитывать дата-сайентист уровня Junior?

In [120]:
# текст запроса
query7_2 = f'''select
                v.name vacancy,
                round(avg (coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to, 0 ))) average_salary
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                and (v.name ilike '%junior%' or v.experience ilike '%Нет опыта%' or v.employment ilike '%Стажировка%')
                group by v.name
                order by average_salary desc 
                limit 5
                '''

In [121]:
# результат запроса
df7_2 = pd.read_sql_query(query7_2, connection)
df7_2

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\246647448.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df7_2 = pd.read_sql_query(query7_2, connection)


vacancy  average_salary
0  Исследователь в области машинного обучения и м...        115000.0
1  Математик-аналитик (data scientist / data anal...         98000.0
2  Программист Python/ Data Scientist (junior / m...         80000.0
3                            Data Scientist (Junior)         72500.0
4          Data Scientist (Computer Vision) / Junior         70000.0

Популярность языков Java, Python и С++ в требованиях работодателей к DS

In [122]:
# текст запроса
query7_3 = f'''select
                'Вакансий для DS (key_skills: Python)' query,
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                and (v.key_skills like '%Python%')
                
                union
                select
                'Вакансий для DS (key_skills: Java)' query,
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                and (v.key_skills like '%Java%')
                
                union
                select
                'Вакансий для DS (key_skills: C++)' query,
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                and (v.key_skills like '%C++%')
                order by cnt_vacancies desc
                
                '''

In [123]:
# результат запроса
df7_3 = pd.read_sql_query(query7_3, connection)
df7_3

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\944850107.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df7_3 = pd.read_sql_query(query7_3, connection)


query  cnt_vacancies
0  Вакансий для DS (key_skills: Python)            351
1     Вакансий для DS (key_skills: C++)             43
2    Вакансий для DS (key_skills: Java)             28

Какое количество вакансий на должность дата-сайентист? 

In [124]:
# текст запроса
query7_4 = f'''select
                'Количество вакансий на должность DS' query,
                count(v.name) cnt_vacancies
                from public.vacancies v
                where (v.name ilike '%data scientist%' or v.name ilike '%data science%'
                or v.name ilike '%исследователь данных%' or v.name ilike '%machine learning%' 
                or v.name ilike '%машинн%обучен%' or (v.name like '%ML%' AND v.name not like '%HTML%')) 
                '''


In [125]:
# результат запроса
df7_4 = pd.read_sql_query(query7_4, connection)
df7_4

C:\Users\Maksym\AppData\Local\Temp\ipykernel_16608\189768922.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df7_4 = pd.read_sql_query(query7_4, connection)


query  cnt_vacancies
0  Количество вакансий на должность DS            480

### Подведем итог исследования, обобщим выводы
- ТОП 5 работодателей по количеству вакансии: 1) Яндекс - 1933, 2) Ростелеком	491, 3)	Тинькофф - 444, 4) СБЕР	- 428, 5) Газпром нефть	- 331.
- Разработкой программного обеспечения занимаются 3553 (15% от общего количества).
- ТОП 3 компании у которых указано 4 сферы деятельности: 101 Интернет, 21vek.by, 2ГИС.
- У 35.8% (8419) компаний не указана сфера деятельности.
- В Москве расположено наибольшее количество работодателей - 2320 и количество вакансий -	5333. 
- Топ 5 работодателя у которых представлены вакансии в наибольшем количестве регионов: 1) Яндекс - 181 регион , 2) Ростелеком	- 152 регион 3) Спецремонт - 116 регион, 4) Поляков Денис Иванович - 88 регион, 5) ООО ЕФИН - 71 регион.
- Компания Яндекс предоставляет свои вакансии в всех городах миллионниках Росии. Общее количество 485 вакансий. 
- Количество вакансий имеющих отношения к данным составляет 1771, 3,5% от общего количества. 
- Среди них на должность DataScience  - 480, что составляет 27%. А также уровня junior: 51 вакансия.
- Количество вакансий на должность DataScience со знаниями (SQl/postgres): 201, со знаниями Python: 351 вакансия.
- Среднее количество необходимых навыков для трудоустройства на должность DataScience: 6,41. Знание (SQl/postgres): 201(41%) и Python - 351 (73%), менее популярно C++ (9%) Java (5,8%)
- DS уровня Junior средняя ЗП 70 тыс - 100 тыс. руб. С опытом работы от 1-го года да 3 лет, средняя ЗП увеличивается в 2 раза,  а с опытом ОТ 3 ДО 6 ЛЕТ: 243 тыс. рублей. 


